### ISSUE TITLE
DoS possibility using Poplar under heavy hitter mode

### ISSUE THREAD

**Issue by SulemanAhmadd** [ietf-wg-ppm/draft-ietf-ppm-dap#413]
The computation required by Poplar VDAF (for heavy hitter computation) is dependent on the number of bits in the measurement value. The bigger the measurement size, the more computation is needed by Poplar to create an aggregate results (as the number of rounds for aggregation increases superlinearly).

In a Sybil attack, malicious clients can craft dummy inputs by setting a very large measurement size which can cause heavy hitters computation to exhibit worst-case performance. This likely results in DoS vector for the Aggregation servers.

**Comment by cjpatton** [comment]
Thanks Suleman! I think this needs a bit more context. Poplar1 requires each measurement to have same fixed length, which all parties (including the Clients) need to agree on prior to VDAF execution. Can you clarify what you mean by "very large"? No measurement size can exceed the fixed length.

**Comment by SulemanAhmadd** [comment]
When deciding the value of "fixed length" for the measurements, care must be taken that DAP deployment can handle heavy hitter computation using Poplar1. The larger the configured length of the measurement, the more computation is required.

I'll expand on by giving an example:

Consider the scenario when we use variable length domain names as the measurement values for Poplar1. We can set the length of each measurement to be fixed at 255 bytes (255 is the maximum number of characters allowed in a domain name by DNS) so that each domain name can fit inside the measurement value.

The domain name distribution has a long tail i.e. only a small number of domains are more than 100 characters. This means that for most cases we would have to encode the domain name and add some sort of padding to reach the 255 bytes fixed measurement size.

When computing the heavy hitter for a batch of such measurements, an optimization we can add is to stop processing (or traversing the IDPF tree) when recovering a measurement if we have reached the padding region. Since, we know what the next set of values are, we do not need to traverse the entire IDPF tree for the computation. This has the implication that for smaller length domain names, It significantly reduces the cost of communication between aggregators and the steps required to recover a heavy hitter measurement.

On the other side, in case of Sybil attacks, malicious client can always use the maximum size domain names forcing the system to always operate on worst-case performance (the optimization cannot be applied) which can potentially hint towards a DoS vector.

---

Basically, the choice of length of measurement directly impacts the required computation required by a DAP deployment superlinearly when using Poplar1 VDAF. This might be more of a consideration for deployment rather a bug in the model.

**Comment by cjpatton** [comment]
Thanks, this makes sense to me! I suggest we add some text in security considerations about this. I don't know if we yet have a DoS risk category, but probably not. Would you mind sending a PR?

**Comment by SulemanAhmadd** [comment]
Sure. I'll create a PR.

**Comment by tgeoghegan** [comment]
#415 was long since merged, so I don't think there's anything left to do in this issue. Thanks again for flagging this problem and making a PR, Suleman!
